bach/bach/dataframe.py

```diff
@@ -2624,6 +2625,112 @@ def dropna(
 
         return dropna_df
 
+    def fillna(
+        self,
+        *,
+        value: Union[Union['Series', Scalar], Dict[str, Union[Scalar, 'Series']]] = None,
+        method: Optional[str] = None,
+        axis: int = 0,
+        inplace: bool = False,
+        limit: Optional[int] = None,
+        sort_by: Optional[Union[str, Sequence[str]]] = None,
+        ascending: Union[bool, List[bool]] = True,
+    ) -> Optional['DataFrame']:
+
+        df = self.copy()
+        if method and value is not None:
+            raise ValueError('cannot specify both "method" and "value".')
+
+        if method in ('ffill', 'pad'):
+            df.ffill(sort_by=sort_by, ascending=ascending, inplace=True)
+
+        elif method in ('bfill', 'backfill'):
+            df.bfill(sort_by=sort_by, ascending=ascending, inplace=True)
+
+        if value is not None:
+            series_to_fill = list(value.keys()) if isinstance(value, dict) else self.data_columns
+            for s in series_to_fill:
+                fill_with = value[s] if isinstance(value, dict) else value
+                df[s] = df.all_series[s].fillna(fill_with)
+
+        if not inplace:
+            return df
+
+        self._update_self_from_df(df)
+        return None
+
+    def ffill(
+        self,
+        sort_by: Optional[Union[str, Sequence[str]]] = None,
+        ascending: Union[bool, List[bool]] = True,
+        inplace: bool = False,
+    ) -> Optional['DataFrame']:
+        from bach.partitioning import Window
+
+        df = self.copy()
+
+        if sort_by:
+            df = df.sort_values(by=sort_by, ascending=ascending)
+
+        if not sort_by and not df.order_by:
+            raise Exception('dataframe must be sorted in order to apply forward fill.')
+
+        # create partition per each non-nullable value in all columns to be interpolated
+        for series_name in self.data_columns:
+            partition_name = f'__partition_{series_name}'
+            df[partition_name] = self.all_series[series_name].copy_override(
+                expression=Expression.construct(
+                    f'case when {{}} is null then 0 else 1 end', self.all_series[series_name],
+                ),
+                name=partition_name,
+            )
+
+            df[partition_name] = df.all_series[partition_name].window_sum(
+                window=Window([], order_by=df.order_by),
+            )
+
+        df.materialize(inplace=True)
```

Comment:
What happens here is quite smart/magical. I needed to look at the SQL generated to actually understand:
in pseudo-sql:
```sql
sum(case when "series_name" is null then 0 else 1 end) over ( order by "df.order_by" asc RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) as "__partition_series_name"
```
I do think it needs a very detailed explanation here in the code.

Reply:
Please give name to the materialization btw
```suggestion
        df.materialize(node_name='fillna_partitioning', inplace=True)
```